borsh-derive-internal/src/generics.rs

```diff
@@ -0,0 +1,228 @@
+    relevant_type_params: HashSet<Ident>,
+
+    // Fields whose type is an associated type of one of the generic type
+    // parameters.
+    associated_type_usage: Vec<&'ast TypePath>,
+}
+
+fn ungroup(mut ty: &Type) -> &Type {
+    while let Type::Group(group) = ty {
+        ty = &group.elem;
+    }
+    ty
+}
+
+impl<'ast> FindTyParams<'ast> {
+    pub fn new(generics: &Generics) -> Self {
+        let all_type_params = generics
+            .type_params()
+            .map(|param| param.ident.clone())
+            .collect();
+
+        FindTyParams {
+            all_type_params,
+            relevant_type_params: HashSet::new(),
+            associated_type_usage: Vec::new(),
+        }
+    }
+    pub fn process(self) -> Vec<TypePath> {
+        let relevant_type_params = self.relevant_type_params;
+        let associated_type_usage = self.associated_type_usage;
+        let mut new_predicates: Vec<TypePath> = relevant_type_params
+            .into_iter()
+            .map(|id| TypePath {
+                qself: None,
+                path: id.into(),
+            })
+            .chain(associated_type_usage.into_iter().cloned())
+            .collect();
+        new_predicates.sort_by_key(|type_path| type_path.to_token_stream().to_string());
+        new_predicates
+    }
+}
+impl<'ast> FindTyParams<'ast> {
+    pub fn visit_field(&mut self, field: &'ast Field) {
+        if let Type::Path(ty) = ungroup(&field.ty) {
+            if let Some(Pair::Punctuated(t, _)) = ty.path.segments.pairs().next() {
+                if self.all_type_params.contains(&t.ident) {
+                    self.associated_type_usage.push(ty);
+                }
+            }
+        }
+        self.visit_type(&field.ty);
+    }
+
+    fn visit_return_type(&mut self, return_type: &'ast ReturnType) {
+        match return_type {
+            ReturnType::Default => {}
+            ReturnType::Type(_, output) => self.visit_type(output),
+        }
+    }
+
+    fn visit_path_segment(&mut self, segment: &'ast PathSegment) {
+        self.visit_path_arguments(&segment.arguments);
+    }
+
+    fn visit_path_arguments(&mut self, arguments: &'ast PathArguments) {
+        match arguments {
+            PathArguments::None => {}
+            PathArguments::AngleBracketed(arguments) => {
+                for arg in &arguments.args {
+                    match arg {
+                        GenericArgument::Type(arg) => self.visit_type(arg),
+                        GenericArgument::AssocType(arg) => self.visit_type(&arg.ty),
+                        GenericArgument::Lifetime(_)
+                        | GenericArgument::Const(_)
+                        | GenericArgument::AssocConst(_)
+                        | GenericArgument::Constraint(_) => {}
+
+                        #[cfg_attr(
+                            feature = "force_exhaustive_checks",
+                            deny(non_exhaustive_omitted_patterns)
```

Comment:
I feel this should be the default configuration for the module of even crate, and explicitly opted-out (allowed) whenever necessary. What do you think?

Reply:
i think that it's so in selene's code (where i borrowed it from) due to convention to specify exact nightly features that are going to be used 
(probably because some combinations of them can interact/bug in weird ways, and it's hard to test all possible combinations in nightly). 
My suggestion, is when `feature(non_exhaustive_omitted_patterns_lint)` is stabilized, then all of `#[cfg_attr(...)]` relevant to it can be removed

Reply:
If the attribute is removed now, then it becomes less apparent, that the feature is not stable yet.